engines/queen/metaengine.cpp

```
@@ -72,8 +78,68 @@ class QueenMetaEngine : public AdvancedMetaEngine<Queen::QueenGameDescription> {
 	int getMaximumSaveSlot() const override { return 99; }
 	bool removeSaveState(const char *target, int slot) const override;
 	int getAutosaveSlot() const override { return 99; }
+
+	void languageInput(Common::Keymap *perLanguage,const char *_commandKeys);
+	Common::KeymapArray initKeymaps(const char *target) const override;
+
 };
 
+const char *const _commandKeys[] = {
+	"ocmglptu", // English
+	"osbgpnre", // German
+	"ofdnepau", // French
+	"acsdgpqu", // Italian
+	"ocmglptu", // Hebrew
+	"acodmthu"  // Spanish
+};
+
+void QueenMetaEngine::languageInput(Common::Keymap *perLanguage, const char* _commandKeys) {
+
+	using namespace Common;
+	using namespace Queen;
+	Action *act;
+
+	act = new Action("OPEN", _("OPEN"));
+	act->setCustomEngineActionEvent(VERB_OPEN);
+	act->addDefaultInputMapping(String(_commandKeys[0]));
```

From backends/keymapper/action.h:
	 * Add a default input mapping for the action
	 *
	 * Unknown hardware inputs will be silently ignored.
	 * Having keyboard bindings by default will not cause trouble
	 * on devices without a keyboard.

Valid hardware input string names are found in backends/keymapper/hardware-input.cpp

---

I see - there's some confusion here as you named the function parameter with an underscore, which matches global name.
I'm still not sure I would recommend a separate keymap per language when the only thing changing between each is the default input mapping. Is language is detected though a game resource that we would know here or is it selectable?

---

The game language is detected by the game resources. I tried to make a function which detect game language and changes the keymap according to the language. But the game language is detected when the game opens. Due to this, I get a null pointer error whenever I try to open the Game Options in the Main UI.